pkg/framework/coremodel/entityframework.cue

```diff
@@ -0,0 +1,99 @@
+package entity
+
+// THIS FILE IS AN EXPERIMENTAL THOUGHT EXERCISE. NAMES AND RULES ARE SUBJECT TO CHANGE.
+// AND DISCUSSION IS ENCOURAGED. IT IS NOT USED AND DOES NOT IMPACT ANY GRAFANA BEHAVIOR.
+//
+// The only things we _really_ know for sure are:
+// - the name "coremodel" is terrible and needs to go away
+// - grafana is moving to schema-centric development
+// - schema-centric development entails things like a consistent API,
+//   storage, and docs framework and/or codegen
+// - if grafana is to have a consistent framework for types that it accepts and
+//   stores via an API, it must also support non-schematized types (raw files)
+// - THESE THINGS NEED PRECISE NAMES OR WE WILL ALL LOSE OUR DAMN MINDS
+// - all of this will only be successful if we work out an iterative transitional
+//   path - no big bang!
+//
+// This file is committed because:
+// - the functional role it will eventually play is defined in a design doc https://bit.ly/3McWhvB
+// - committing a skeleton like this is an iterative step towards actually implementing ^
+// - @sdboyer keeps recreating this ad hoc in 1:1s to illustrate the concept
+
+
+
+// An EntityType defines a type in Grafana's entity system.
+//
+// An entity is a sequence of bytes - for example, a JSON file or HTTP request
+// body - that conforms to the constraints defined in an EntityType, and
+// enforced by Grafana's entity system.
+//
+// Once Grafana's entity system has determined a given byte sequence to be an
+// instance of a known EntityType, type-specific behaviors can be applied,
+// requests can be routed, events can be fired, etc.
+//
+// Classes and objects in most programming languages are analogous:
+//   - EntityType is like a `class` keyword
+//   - Each declaration of #EntityType is like a class declarations
+//   - Byte sequences are like arguments to the class's constructor
+//   - Entities are like objects - what's returned from the constructor
+//
+// There are two fundamental kinds of EntityTypes: RawEntityType and StructuredEntityType.
+#EntityType: #RawEntityType | #StructuredEntityType
```

Comment:
I am struggling to understand why is it called `EntityType` rather than `EntityKind` - I thought we settled on `Kind` for the  e.g. from the doc:

> Coremodel: A collection of versioned schemas for a single “kind” of object, the logic for translating between those versions, and metadata on how they are to be used.


`type` would be reserved for grouping within a `kind`, e.g. the current datasource type

Reply:
i'm not terribly strongly attached to the use of kind vs. type here. i had written it at least three different ways before putting this PR up, sometimes including "kind" at the top level instead. i could be convinced either way. Though -

> `type` would be reserved for grouping within a `kind`, e.g. the current datasource type

This linkage raises yellow flags for me, because i think of these things in terms of schema compositions, and there is no _necessary_ limit on the number of things that might be composed.

For example, today the base datasource schema (whenever we write it) certainly composes a datasource plugin, so it makes sense to think of that as a simple parent/child typing relation.

Panels, by contrast, compose two things - query from datasource plugins, and panel from panel plugins. We conventionally talk about them in terms of the latter (we'd say `type: panel, kind: heatmap` in your wording here), but IMO there's no clear, necessary reason that's right - it would be just as correct to describe them in terms of their datasource.

There's also an interesting, not-implausible scenario for datasources as well - there are a few flavors of prometheus (e.g. vanilla vs. with sigv4) that might want us to allow _further_ differentiation of a `prometheus`-typed ds kind beyond that single level. Alternatively, we might want `type` of a datasource to refer to the query language it speaks, rather than the specific plugin.

All of these make me concerned that making a hard assumption about a 'type-subtype' relation between `kind` and `type` may bite us.

Reply:
any alternatives?  we currently just use "type" for everything.  Or rather we way "a panel of type heatmap" or "a datasource of type influx" -- to make it confusing, we also say "a plugin of type influx" and "a transformer of type joinByLabels"

Reply:
> This linkage raises yellow flags for me, because i think of these things in terms of schema compositions, and there is no necessary limit on the number of things that might be composed.

> Panels, by contrast, compose two things - query from datasource plugins, and panel from panel plugins. We conventionally talk about them in terms of the latter (we'd say kind: panel, type: heatmap in your wording here - though it's also important to note that panels are not a top-level thing in the schema system, and we aren't close to being able to make them one), but IMO there's no clear, necessary reason that's right - it would be just as correct to describe them in terms of their datasource.

Hmm... I see your general point of being able to categorize an entity in different ways, but I don't follow this example. The `type` can be useful even if the entity in question made no use of composability, and never referred to other entities. 

Imagine we had datasource concept - we would still want to differentiate between a "text" panel and a "image" panel. This isn't to say there is one-true-way of defining the `type` - we define and use it because having a clear separation between groups of entities with shared characteristics/features is useful from the application PoV.

It's a very interesting point!

Reply:
(Just to be clear, i'm not trying to change the way anyone talks about it. IMO the conversational shorthand is totally a-ok 👍🏻 excelente.)

> any alternatives?

Having a precise alternative is why i invested so much thinking into [composition and "slots"](https://docs.google.com/document/d/12vYU-DRlS0yYxvXW52cOH7kSIxmJAER_4_mlA0An_QM/edit#heading=h.knp2lhkqkje3) (a name i still hate). That should (i hope?) offer precise terms for the whole problem space. So:

The dashboard lineage has `Panel` and `Query` slots, each of which [accepts N conformant lineages](https://github.com/grafana/grafana/blob/v8.5.0/packages/grafana-schema/src/scuemata/dashboard/dist/family.cue#L30-L44). Each panel instance will be an instance of exactly one of the `Panel` and `Query` slot implementations (or maybe zero-or-one, with zero being the case where the plugin has provided no schema).

Therefore, ignoring the issue of panels not being top-level thingies for the moment, a serialized form of this info for a particular instance of a panel in a dashboard might be `panel(Panel:timeseries,Query:prometheus)`.

Reply:
> The type can be useful even if the entity in question made no use of composability, and never referred to other entities.

Sure - of course it "can be useful." But if such expansion is permitted, then the meaning/intended interpretation of the value of `type` becomes arbitrary. That's not a problem on its own - from the system's perspective, schema contents are arbitrary. That's how it should be.

But [this code](https://github.com/grafana/grafana/pull/56507/files#diff-2724413b7c4fd4e0590ba2b13c066a02d273ffc76f2e7c19bc3d87bebf10362cR111), `kind` and `type`  are being given a special, system-level function: generic components of a reference, which generic algorithms in the system are supposed to be able to do _something_ meaningful with. If `type` is just an opaque string, dependent on `kind`, it limits what any generic algorithm can really do. It also forces us to design a considerably more complex system, where instead of defining a generic algorithm, validity and backwards compatibility rules _once_, we do that _for each kind_, and it becomes the system's responsibility to try to enforce meta-rules on what's been delegated to the kinds. (Which is basically impossible, so we won't even try)

IMO, keeping a clean separation between system-level and type-level responsibilities is one of the most important things we can do to avoid [sprawl and incidental complexity](https://github.com/grafana/grafana/pull/56492#issuecomment-1272293604).